crates/federated-server/src/server/graph_fetch_method.rs

```diff
@@ -21,81 +15,60 @@ pub enum GraphFetchMethod {
-        hot_reload_config_path: Option<PathBuf>,
-        sender: watch::Sender<Option<Arc<Engine<GatewayRuntime>>>>,
-        hooks: HooksWasi,
-    ) -> crate::Result<()> {
+    /// Converts the fetch method into a stream of graph definitions.
+    ///
+    /// This can happen in two ways: if providing a graph SDL, we return a new graph immediately.
+    /// Alternatively, if a graph ref and access token is provided, the function returns
+    /// immediately, and runs a background process to fetch the graph definition from the GDN
+    pub(crate) async fn into_stream(self) -> crate::Result<GraphStream> {
         #[cfg(feature = "lambda")]
         if matches!(self, GraphFetchMethod::FromGraphRef { .. }) {
             return Err(crate::Error::InternalError(
                 "Cannot fetch schema with graph in lambda mode.".to_string(),
             ));
         }
 
+        let (sender, receiver) = mpsc::channel(4);
+
         match self {
             GraphFetchMethod::FromGraphRef {
                 access_token,
                 graph_ref,
             } => {
-                let config = config.clone();
                 tokio::spawn(async move {
-                    let config = config.clone();
                     use super::graph_updater::GraphUpdater;
 
-                    GraphUpdater::new(graph_ref, access_token, sender, config, hooks)?
-                        .poll()
-                        .await;
+                    GraphUpdater::new(graph_ref, access_token, sender)?.poll().await;
 
                     Ok::<_, crate::Error>(())
                 });
             }
-            GraphFetchMethod::FromSchema {
-                federated_sdl,
-                reload_signal,
-            } => {
-                let gateway = gateway::generate(
-                    GraphDefinition::Sdl(federated_sdl),
-                    config,
-                    hot_reload_config_path.clone(),
-                    hooks.clone(),
-                )
-                .await?;
-
-                sender.send(Some(Arc::new(gateway)))?;
+            GraphFetchMethod::FromSchema { federated_sdl } => {
+                sender.send(GraphDefinition::Sdl(federated_sdl)).await?;
 
-                if let Some(reload_signal) = reload_signal {
-                    tokio::spawn(async move {
-                        let mut stream = ReceiverStream::new(reload_signal);
-
-                        while let Some((sdl, config)) = stream.next().await {
-                            let gateway = gateway::generate(
-                                GraphDefinition::Sdl(sdl),
-                                &config,
-                                hot_reload_config_path.clone(),
-                                hooks.clone(),
-                            )
-                            .await?;
-                            sender.send(Some(Arc::new(gateway)))?;
+                // Leak the sender so the channel never closes
+                Box::leak(Box::new(sender));
```

Comment:
Same nitpick, if you leak would say that this function is only called once

Reply:
Yeah, good point.  Although I wonder if I'd actually be better off doing this without leaking, just incase someone decides to call this multiple times...